File: src/main/scala/org/raisercostin/jedi/impl/FileVisitor.scala
package org.raisercostin.jedi.impl

import java.nio.file._
import java.nio.file.attribute.BasicFileAttributes
import scala.collection.Traversable
import java.io.File

class TraversePath(path: Path, withDir: Boolean = false) extends Traversable[(Path, BasicFileAttributes)] {
  override def foreach[U](f: ((Path, BasicFileAttributes)) => U) {
    class Visitor extends SimpleFileVisitor[Path] {
      override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
        //println("visited "+file)
        f(file -> attrs)
        FileVisitResult.CONTINUE
        //} catch {
        //case _: Throwable => FileVisitResult.TERMINATE
      }
      override def preVisitDirectory(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
        if (withDir) {
          //println("visited folder "+file)
          f(file -> attrs)
        }
        FileVisitResult.CONTINUE
      }
    }
    Files.walkFileTree(path, new Visitor)
  }
}

trait FileVisitor {
  def apply(path: String) = fromNioPath(new File(path).toPath)
  def apply(path: Path) = fromNioPath(path)

  import scala.language.implicitConversions
  implicit def fromNioPath(path: Path): TraversePath = new TraversePath(path)
  implicit def fromIoFile(file: File): TraversePath = new TraversePath(file.toPath)
}
object FileVisitor extends FileVisitor

File: src/main/kotlin/org/raisercostin/jedi/impl/FileVisitor.kt
package org.raisercostin.jedi.impl

import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import scala.collection.Traversable
import java.io.File

class TraversePath(path: Path, ,Dir: Boolean = false) : Traversable<(Path, BasicFileAttributes)> {
  override fun foreach<U>(f: ((Path, BasicFileAttributes)) -> U) {
    class Visitor : SimpleFileVisitor<Path> {
      override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
        f(file -> attrs)
        FileVisitResult.CONTINUE
        //} catch {
        //else: Throwable -> FileVisitResult.TERMINATE
      }
      override fun preVisitDirectory(file: Path, attrs: BasicFileAttributes): FileVisitResult {
        if (,Dir) {
          f(file -> attrs)
        }
        FileVisitResult.CONTINUE
      }
    }
    Files.walkFileTree(path, Visitor)
  }
}

interface FileVisitor {
  fun apply(path: String) = fromNioPath(new File(path).toPath)
  fun apply(path: Path) = fromNioPath(path)

  import scala.language.implicitConversions
  implicit fun fromNioPath(path: Path): TraversePath = TraversePath(path)
  implicit fun fromIoFile(file: File): TraversePath = TraversePath(file.toPath)
}
object FileVisitor : FileVisitor